docs/architecture-decision-records/20220121-conduit-plugin-architecture.md

```diff
@@ -0,0 +1,322 @@
+- `Start` - Conduit signals to the plugin it wants it to start running. The plugin is expected to open any connections
+  needed to write records. In case of a successful response, the next call will be `Run`.
+- `Run` - this function opens a bidirectional stream between Conduit and the plugin, same as in the source plugin,
+  except that in this case Conduit is sending records to the plugin and receiving acknowledgments back. Both streams are
+  independent and are able to transmit data concurrently. The plugin is expected to send an acknowledgment back to
+  Conduit for every record it received, even if the record was not processed successfully (in that case the
+  acknowledgment should contain the error). The stream should stay open until either an error occurs or the `Stop`
+  function is called *and* all remaining acknowledgments are sent (this is handled by the SDK).
+- `Stop` - Conduit signals to the plugin that there be no more records written to the request stream in `Run`. The
+  plugin needs to flush any records that are cached in memory, send back the acknowledgments and stop the `Run`
+  function.
+- `Teardown` - see the description of `Teardown` for the source plugin (same purpose).
+
+Last but not least, the specifier interface:
+
+```protobuf
+service SpecifierPlugin {
+  rpc Specify(Specifier.Specify.Request) returns (Specifier.Specify.Response);
+}
+```
+
+This one is very simple and allows the plugin to describe itself. Through this Conduit will get to know the name of the
+plugin, version, author, description and data about what parameters the plugin expects in the configuration.
+
+### Conduit internal plugin interface
+
+Conduit internally defines its own plugin interfaces:
+
+```go
+type SourcePlugin interface {
+  Configure(context.Context, map[string]string) error
+  Start(context.Context, record.Position) error
+  Read(context.Context) (record.Record, error)
+  Ack(context.Context, record.Position) error
+  Stop(context.Context) error
+  Teardown(context.Context) error
+}
+
+type DestinationPlugin interface {
+  Configure(context.Context, map[string]string) error
+  Start(context.Context) error
+  Write(context.Context, record.Record) error
+  Ack(context.Context) (record.Position, error)
+  Stop(context.Context) error
+  Teardown(context.Context) error
+}
+
+type SpecifierPlugin interface {
+  Specify() (Specification, error)
+}
+```
+
+The methods roughly map to the methods defined in the gRPC interface described in the previous chapter. There are 2
+things to point out
+
+- These interfaces use internal structs in the Conduit codebase (e.g. `record.Position`, `record.Record`), so
+  interacting with a plugin is seamless.
+- The bidirectional stream method `Run` is broken down into two separate methods, one for receiving messages from the
+  stream and one to send messages. Those methods can only be called after a call to `Start` since that is the method in
```

Comment:
I see it differently - the sentence here is not addressing any specific plugin (i.e. source or destination), but talking generally about the `Run` method in both plugins. We can't really say that we are sending acknowledgments, because that would be true for the source plugin, but not for the destination, since there we are sending records and receiving acknowledgments.

Reply:
Ah, my bad, indeed no specific plugin is mentioned.